src/pmem.c

```diff
@@ -602,17 +624,207 @@ memmove_nodrain_normal(void *pmemdest, const void *src, size_t len)
 	return memmove(pmemdest, src, len);
 
 }
+
+/*
+ * memcpy_nodrain_normal -- (internal) memmcpy to pmem without hw drain
+ */
+static void *
+memcpy_nodrain_normal(void *pmemdest, const void *src, size_t len)
+{
+	return memcpy(pmemdest, src, len);
+}
+
+/*
+ * nodrain_movnt -- (internal) memXXX worker function without hw drain, movnt
+ */
+
+static void *
+nodrain_movnt(void *pmemdest, void *src, size_t len)
+{
+	/* way too chatty for LOG level 3 */
+	LOG(15, "pmemdest %p src %p len %zu", pmemdest, src, len);
+
+	__m128i xmm0, xmm1, xmm2, xmm3, xmm4, xmm5, xmm6, xmm7;
+	size_t cnt = 0;
+	uint32_t i = 0;
+	__m128i *d;
+	__m128i *s;
+	void *tempdest = pmemdest;
+
+	if (len == 0)
+		return NULL;
+
+	if (src == pmemdest)
+		return pmemdest;
+	/*
+	 * Calculate the number of bytes to move forward to get
+	 * to the nearest 16 byte boundary.
+	 */
+	if ((cnt = ((uint64_t)pmemdest & MOVNT_MASK)) != 0)
+		cnt = MOVNT_SIZE - cnt;
+	if (cnt != 0) {
+		/* dest is unaligned */
+		uint8_t	*d8 = (uint8_t *)pmemdest;
+		const uint8_t *s8 = (uint8_t *)src;
+		for (i = 0; i < cnt; i++) {
+			*d8 = *s8;
+			d8++;
+			s8++;
+		}
+		pmemdest += cnt;
+		src += cnt;
+		len -= cnt;
+		pmem_flush(d8, cnt);
+	}
+
+	d = (__m128i *)pmemdest;
+	s = (__m128i *)src;
+
+	cnt = len >> CHUNK_SHIFT;
+	for (i = 0; i < cnt; i++) {
+		xmm0 = _mm_loadu_si128(s);
+		xmm1 = _mm_loadu_si128(s + 1);
+		xmm2 = _mm_loadu_si128(s + 2);
+		xmm3 = _mm_loadu_si128(s + 3);
+		xmm4 = _mm_loadu_si128(s + 4);
+		xmm5 = _mm_loadu_si128(s + 5);
+		xmm6 = _mm_loadu_si128(s + 6);
+		xmm7 = _mm_loadu_si128(s + 7);
+		s += 8;
+		/* Store double quadword with non-temporal hint */
+		_mm_stream_si128(d,	xmm0);
+		_mm_stream_si128(d + 1,	xmm1);
+		_mm_stream_si128(d + 2,	xmm2);
+		_mm_stream_si128(d + 3,	xmm3);
+		_mm_stream_si128(d + 4,	xmm4);
+		_mm_stream_si128(d + 5,	xmm5);
+		_mm_stream_si128(d + 6,	xmm6);
+		_mm_stream_si128(d + 7,	xmm7);
+		d += 8;
+	}
+
+	/* Copy the tail (<128 bytes) in 16 bytes chunks */
+	len &= CHUNK_MASK;
+	if (len != 0) {
+		cnt = len >> MOVNT_SHIFT;
+		for (i = 0; i < cnt; i++) {
+			xmm0 = _mm_loadu_si128(s);
+			_mm_stream_si128(d, xmm0);
+			s++;
+			d++;
+		}
+	}
+
+	/* Copy the last bytes (<16). First dwords then bytes */
+	len &= MOVNT_MASK;
+	if (len != 0) {
+		cnt = len >> DWORD_SHIFT;
+		int32_t	*d32 = (int32_t *)d;
+		int32_t	*s32 = (int32_t *)s;
+
+		for (i = 0; i < cnt; i++) {
+			_mm_stream_si32(d32, *s32);
+			d32++;
+			s32++;
+		}
+
+		cnt =	len & DWORD_MASK;
+		uint8_t	*d8 = (uint8_t *)d32;
+		const uint8_t	*s8 = (uint8_t *)s32;
+
+		for (i = 0; i < cnt; i++) {
+			*d8 = *s8;
+			d8++;
+			s8++;
+		}
+		pmem_flush(d8, cnt);
+	}
+	return tempdest;
+}
+
 /*
  * memmove_nodrain_movnt -- (internal) memmove to pmem without hw drain, movnt
  */
 static void *
 memmove_nodrain_movnt(void *pmemdest, const void *src, size_t len)
 {
+
+	void	*src1, *dest1;
+	int	i;
+	uint64_t chunk_size = util_nonoverlap_range((void *)src, pmemdest, len);
```

Comment:
Rather than *const-cast* on `src`, modify the `util_nonoverlap_range` prototype.